### PR TITLE
Multilingual workflow

### PR DESCRIPTION
**Pull request by wiewiurdp**
# Description
 
Multilingual workflow

Issue #803 
## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

- [x] Unit Tests
- [x] Behat Test

# Checklist:
- [x] I have read the contribution requirements and fulfill them.
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: module/product/migrations/Version20180619083830.php [hunk 1]
```diff
@@ -112,13 +112,30 @@ public function up(Schema $schema): void
...
+            'INSERT INTO privileges (id, code, area) VALUES (?, ?, ?)',
+            [Uuid::uuid4()->toString(), 'PRODUCT_READ', 'Product']
+        );
+        $this->addSql(
+            'INSERT INTO privileges (id, code, area) VALUES (?, ?, ?)',
+            [Uuid::uuid4()->toString(), 'PRODUCT_UPDATE', 'Product']
+        );
+        $this->addSql(
+            'INSERT INTO privileges (id, code, area) VALUES (?, ?, ?)',
+            [Uuid::uuid4()->toString(), 'PRODUCT_DELETE', 'Product']
```
Comment (wfajczyk) [inline module/product/migrations/Version20180619083830.php:138]: Why the function `createProductPrivileges` is not use?

File: module/workflow/src/Application/Controller/Api/Transition/TransitionCreateAction.php [hunk 2]
```diff
@@ -86,16 +86,14 @@ public function __construct(CommandBusInterface $commandBus, FormFactoryInterfac
      *     @SWG\Schema(ref="#/definitions/validation_error_response")
      * )
      *
-     * @ParamConverter(class="Ergonode\Workflow\Domain\Entity\Workflow")
-     *
-     * @param Workflow $workflow
-     * @param Request  $request
+     * @param AbstractWorkflow $workflow
+     * @param Request          $request
```
Comment (wfajczyk) [inline module/workflow/src/Application/Controller/Api/Transition/TransitionCreateAction.php:90]: This working when You drop `ParamConverter` ?

File: module/workflow/src/Application/Controller/Api/Transition/TransitionDeleteAction.php [hunk 3]
```diff
@@ -83,17 +83,16 @@ public function __construct(CommandBusInterface $commandBus)
      * @ParamConverter(class="Ergonode\Workflow\Domain\Entity\Status", name="source")
      * @ParamConverter(class="Ergonode\Workflow\Domain\Entity\Status", name="destination")
      *
-     * @param Workflow $workflow
-     * @param Status   $source
-     * @param Status   $destination
+     * @param AbstractWorkflow $workflow
+     * @param Status           $source
+     * @param Status           $destination
```
Comment (wfajczyk) [inline module/workflow/src/Application/Controller/Api/Transition/TransitionDeleteAction.php:88]: I don't understand:
double `ParamConverter` to `Status` and no  `ParamConverter` to `AbstractWorkflow` 

I think about params `Status` is to delete

File: module/workflow/src/Application/Controller/Api/Workflow/WorkflowCreateAction.php [hunk 4]
```diff
@@ -106,24 +106,24 @@ public function __construct(
      */
     public function __invoke(Request $request): Response
     {
-        $data = $request->request->all();
+        $type = $request->request->get('type');
```
Comment (wfajczyk) [inline module/workflow/src/Application/Controller/Api/Workflow/WorkflowCreateAction.php:109]: What if the parameter `type` is missing

File: module/workflow/src/Application/Controller/Api/Workflow/WorkflowTypeDictionaryAction.php [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+    }
+
+    /**
+     * @SWG\Tag(name="Dictionary")
+     * @SWG\Parameter(
+     *     name="language",
+     *     in="path",
+     *     type="string",
+     *     description="Language code",
+     *     default="en"
```
Comment (wfajczyk) [inline module/workflow/src/Application/Controller/Api/Workflow/WorkflowTypeDictionaryAction.php:55]:
```suggestion
     *     default="en_GB"
```

File: module/workflow/src/Application/Controller/Api/Workflow/WorkflowTypeDictionaryAction.php [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+     * )
+     * @SWG\Response(
+     *     response=200,
+     *     description="Returns collection supported product types"
```
Comment (wfajczyk) [inline module/workflow/src/Application/Controller/Api/Workflow/WorkflowTypeDictionaryAction.php:59]: `supported product types` ?

File: module/workflow/src/Infrastructure/Grid/Builder/Query/StatusAttributeDataSetQueryBuilder.php [hunk 7]
```diff
@@ -36,14 +36,10 @@ public function addSelect(QueryBuilder $query, string $key, AbstractAttribute $a
     {
         $query->addSelect(sprintf(
             '(
-                SELECT value FROM product_value pv
-                JOIN value_translation vt ON vt.value_id = pv.value_id
-                WHERE pv.attribute_id = \'%s\'
-                AND pv.product_id = p.id
-                AND vt.language IS NULL
+                SELECT status_id FROM product_workflow_status pws
+                WHERE pws.product_id = p.id
```
Comment (wfajczyk) [inline module/workflow/src/Infrastructure/Grid/Builder/Query/StatusAttributeDataSetQueryBuilder.php:40]: Is `WHERE` should  `Language` param ?

**Review by wfajczyk — CHANGES_REQUESTED** [review]

**Review by wfajczyk — APPROVED** [review]

File: module/product/src/Domain/Entity/AbstractProduct.php [hunk 8]
```diff
@@ -84,21 +84,22 @@ public function __construct(
...
-            $categories,
-            $attributes
         ));
+
+        foreach ($categories as $categoryId) {
+            $this->addToCategory($categoryId);
+        }
+
+        foreach ($attributes as $code => $value) {
+            if ($value) {
```
Comment (BastekBielawski) [inline module/product/src/Domain/Entity/AbstractProduct.php:99]: `array_filter` will cleanup `null` values from `$attributes`. That `if` will be erased

**Review by BastekBielawski — CHANGES_REQUESTED** [review]

File: module/product/src/Infrastructure/Handler/AbstractCreateProductHandler.php [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+    }
+
+    /**
+     * @param array $attributes
+     *
+     * @return array
+     */
+    protected function addAudit(array $attributes): array
+    {
+        $token = $this->tokenStorage->getToken();
```
Comment (piotrkreft) [inline module/product/src/Infrastructure/Handler/AbstractCreateProductHandler.php:91]: You could use `\Symfony\Component\Security\Core\Security` - it's just a nice wrapper to get User

File: module/product/src/Persistence/Dbal/Projector/AbstractProductValueProjector.php [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+    /**
+     * @param string         $productId
+     * @param string         $attributeId
+     * @param ValueInterface $value
+     *
+     * @throws DBALException
+     */
+    protected function insertValue(string $productId, string $attributeId, ValueInterface $value): void
+    {
+        if ($value instanceof StringValue) {
```
[the remark this answers is not in the document]
Comment (piotrkreft) [reply module/product/src/Persistence/Dbal/Projector/AbstractProductValueProjector.php:48]: 👍🏻  if not I guess switch would increase readability

File: module/workflow/src/Application/Controller/Api/Product/ProductWorkflowAction.php [hunk 11]
```diff
@@ -22,15 +22,15 @@
...
+ *     path="products/{product}/workflow/{productLanguage}",
  *     methods={"Get"},
  *     requirements={"product"="[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}"}
  * )
  */
 class ProductWorkflowAction
 {
     /**
-     * @var ProductWorkflowQuery $query;
+     * @var ProductWorkflowQuery $query ;
```
Comment (piotrkreft) [inline module/workflow/src/Application/Controller/Api/Product/ProductWorkflowAction.php:33]: the semicolon is redundant at all

File: module/workflow/src/Application/DependencyInjection/CompilerPass/CreateWorkflowCommandFactoryProviderInterfaceCompilerPass.php [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+use Symfony\Component\DependencyInjection\Compiler\CompilerPassInterface;
+use Symfony\Component\DependencyInjection\ContainerBuilder;
+use Symfony\Component\DependencyInjection\Reference;
+use Ergonode\Workflow\Infrastructure\Provider\CreateWorkflowCommandFactoryProvider;
+
+/**
+ */
+class CreateWorkflowCommandFactoryProviderInterfaceCompilerPass implements CompilerPassInterface
+{
+    public const TAG = 'component.workflow.create_workflow_command_factory_interface';
```
Comment (piotrkreft) [inline module/workflow/src/Application/DependencyInjection/CompilerPass/CreateWorkflowCommandFactoryProviderInterfaceCompilerPass.php:21]: I suggest configuration over logic.
https://github.com/ergonode/backend/blob/develop/module/core/src/Resources/config/services.yml#L53

File: module/workflow/src/Application/DependencyInjection/CompilerPass/UpdateWorkflowCommandFactoryProviderInterfaceCompilerPass.php [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+use Symfony\Component\DependencyInjection\Compiler\CompilerPassInterface;
+use Symfony\Component\DependencyInjection\ContainerBuilder;
+use Symfony\Component\DependencyInjection\Reference;
+use Ergonode\Workflow\Infrastructure\Provider\UpdateWorkflowCommandFactoryProvider;
+
+/**
+ */
+class UpdateWorkflowCommandFactoryProviderInterfaceCompilerPass implements CompilerPassInterface
+{
+    public const TAG = 'component.workflow.update_workflow_command_factory_interface';
```
Comment (piotrkreft) [inline module/workflow/src/Application/DependencyInjection/CompilerPass/UpdateWorkflowCommandFactoryProviderInterfaceCompilerPass.php:21]: As above

File: module/workflow/src/Application/DependencyInjection/CompilerPass/WorkflowFormCompilerPass.php [hunk 14]
```diff
@@ -0,0 +1,48 @@
...
+use Ergonode\Workflow\Application\Provider\WorkflowFormProvider;
+use Symfony\Component\DependencyInjection\Compiler\CompilerPassInterface;
+use Symfony\Component\DependencyInjection\ContainerBuilder;
+use Symfony\Component\DependencyInjection\Reference;
+
+/**
+ */
+class WorkflowFormCompilerPass implements CompilerPassInterface
+{
+    public const TAG = 'workflow.form.workflow_form_interface';
```
Comment (piotrkreft) [inline module/workflow/src/Application/DependencyInjection/CompilerPass/WorkflowFormCompilerPass.php:21]: As above

File: module/workflow/src/Application/DependencyInjection/CompilerPass/WorkflowTypeCompilerPass.php [hunk 15]
```diff
@@ -0,0 +1,48 @@
...
+
+use Symfony\Component\DependencyInjection\Compiler\CompilerPassInterface;
+use Symfony\Component\DependencyInjection\ContainerBuilder;
+use Ergonode\Workflow\Application\Provider\WorkflowTypeProvider;
+
+/**
+ */
+class WorkflowTypeCompilerPass implements CompilerPassInterface
+{
+    public const TAG = 'workflow.domain.workflow_interface';
```
Comment (piotrkreft) [inline module/workflow/src/Application/DependencyInjection/CompilerPass/WorkflowTypeCompilerPass.php:20]: As above

File: module/workflow/src/Application/DependencyInjection/CompilerPass/CreateWorkflowCommandFactoryProviderInterfaceCompilerPass.php (hunk 12 above)
Comment (piotrkreft) [inline module/workflow/src/Application/DependencyInjection/CompilerPass/CreateWorkflowCommandFactoryProviderInterfaceCompilerPass.php:21]: Should just be `component.workflow.create_workflow_command_factory`

File: module/workflow/src/Application/Form/Model/Workflow/WorkflowFormModel.php [hunk 16]
```diff
@@ -0,0 +1,43 @@
...
+     *     max=100,
+     *     maxMessage="Workflow code is to long, It should have {{ limit }} character or less.",
+     *     groups={"Create"}
+     *     )
+     * @WorkflowExists(groups={"Create"})
+     */
+    public ?string $code = null;
+
+    /**
+     * @var array
```
Comment (piotrkreft) [inline module/workflow/src/Application/Form/Model/Workflow/WorkflowFormModel.php:34]: If it's cool with you use `Type[]` please.

File: module/workflow/src/Application/Provider/WorkflowTypeProvider.php [hunk 17]
```diff
@@ -0,0 +1,41 @@
...
+    /**
+     * @param string ...$classes
+     *
+     * @throws \ReflectionException
+     */
+    public function __construct(string ...$classes)
+    {
+        $this->types = [];
+        foreach ($classes as $class) {
+            $type = (new \ReflectionClass($class))->getConstant('TYPE');
```
Comment (piotrkreft) [inline module/workflow/src/Application/Provider/WorkflowTypeProvider.php:29]: Types should be just injected

File: module/workflow/src/Application/Provider/WorkflowTypeProvider.php [hunk 18]
```diff
@@ -0,0 +1,41 @@
...
+            $this->types[$type] = $class;
+        }
+    }
+
+    /**
+     * @return array
```
Comment (piotrkreft) [inline module/workflow/src/Application/Provider/WorkflowTypeProvider.php:35]:
```suggestion
     * @return string[]
```

File: module/workflow/src/Domain/Entity/AbstractWorkflow.php [hunk 19]
```diff
@@ -0,0 +1,429 @@
...
+
+    /**
+     * @param StatusId $id
+     *
+     * @throws \Exception
+     */
+    public function setDefaultStatus(StatusId $id): void
+    {
+        if (!$this->hasStatus($id)) {
+            throw  new \RuntimeException(sprintf('Status "%s" not exists', $id->getValue()));
```
Comment (piotrkreft) [inline module/workflow/src/Domain/Entity/AbstractWorkflow.php:124]: Domain should probably throw Domain exceptions, not global ones

**Review by piotrkreft — CHANGES_REQUESTED** [review]

File: module/workflow/src/Application/Form/Model/Workflow/WorkflowFormModel.php (hunk 16 above)
Comment (wiewiurdp) [reply module/workflow/src/Application/Form/Model/Workflow/WorkflowFormModel.php:34]: this is an array of uuids

File: module/workflow/src/Application/DependencyInjection/CompilerPass/UpdateWorkflowCommandFactoryProviderInterfaceCompilerPass.php (hunk 13 above)
Comment (wiewiurdp) [reply module/workflow/src/Application/DependencyInjection/CompilerPass/UpdateWorkflowCommandFactoryProviderInterfaceCompilerPass.php:21]: As disussed I would try to do it later

File: module/workflow/src/Domain/Entity/AbstractWorkflow.php (hunk 19 above)
Comment (wiewiurdp) [reply module/workflow/src/Domain/Entity/AbstractWorkflow.php:124]: As dissused it need to be talked through on backend side

**Review by piotrkreft — APPROVED** [review]

**Review by wfajczyk — APPROVED** [review]

File: module/product/src/Persistence/Dbal/Projector/AbstractProductValueProjector.php (hunk 10 above)
[the remark this answers is not in the document]
Comment (wiewiurdp) [reply module/product/src/Persistence/Dbal/Projector/AbstractProductValueProjector.php]: discussed with Rafal that it needs to be talk through

**Review by piotrkreft — APPROVED** [review]

**Review by BastekBielawski — APPROVED** [review]

**Pull request merged**